ORCA-109/output_csv

BWMac opened the pull request:
This PR adds the CSV_OUTPUT and SYNAPSE_STORE processes, finishing off v1 of the ome.tiff validation workflow. CSV_OUTPUT takes in a list of all fo the JSON file paths generated by previous processes and aggregates the data, combined with data from the input csv file, to generate the final output csv file. SYNAPSE_STORE takes this csv file and saves it in a synapse folder provided by the user (the default `params.parent_folder` is a public folder that I own for testing). 

Bruno and I had to make some regressive changes to `package_validate.py` and `utils.py` to accommodate python 2 which is installed in the `openmicroscopy/bftools:latest` docker container. This is a temporary solution that will need to be addressed in future versions.

Edit: I realized that there is an issue with the `params.input` path being provided to the CSV_OUTPUT process which causes the workflow to not run on Tower. I am working on a fix, but the workflow will run locally.

BrunoGrandePhD commented on main.nf:
```diff
@@ -156,12 +193,18 @@ workflow {
     def parser = new groovy.json.JsonSlurper()
     return parser.parseText(file.text)
-}
+}
```
Don't forget the empty line at the end of every file. I recommend setting the following preference in VS Code so it's done automatically for you ([source](https://stackoverflow.com/a/44704969)): 

```json
{
    "files.insertFinalNewline": true
}
```

BrunoGrandePhD commented on main.nf:
```diff
@@ -156,12 +193,18 @@ workflow {
+        | XMLVALID_VALIDATE \
+            | mix( SYNAPSE_CHECK.out, MD5_VALIDATE.out, FILE_EXT_VALIDATE.out, SHOWINF_VALIDATE.out, XMLVALID_VALIDATE.out ) \
+            | map { it[2] }
+            | collect \
+            | CSV_OUTPUT
```
Before I forget, this should use the change you introduced:

```suggestion
            | collect \
            | map { tuple(it, params.input) }
            | CSV_OUTPUT
```

BrunoGrandePhD reviewed (changes requested):
Here's my first-pass review. Once the requested changes are made, I'll take another pass and likely will test the workflow itself.

thomasyu888 reviewed (commented):
Looks great!  Some comments on my end too.  Thanks @BrunoGrandePhD for the review.

BWMac commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+                output_dict[syn_id]["failed_tests"].append(test_name)
+            # set is_valid
+            if (
+                output_dict[syn_id]["passed_tests"]
+                == output_dict[syn_id]["required_tests"]
```
the lists are sorted below, does that account for your concern?

thomasyu888 replied:
Right, but right here, you're doing a comparison for if the two lists are equal, if so, the tests passed.  

If the lists are unordered here, then it'll be a "failed" test even though technically the two lists are comprised of the same values.

BWMac replied:
Oh, right. Makes sense

BrunoGrandePhD replied:
We opted to perform the logic using sets. Specifically, the set of passed tests should be a superset of the set of required tests, which is more future-proof.

BWMac commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+                        "is_valid": "",
+                        "passed_tests": [],
+                        "failed_tests": [],
+                        "skipped_tests": "",  # placeholder for future version with skipping
```
I think leaving it as "" is best, or else "[]" ends up in the exported csv file under skipped_tests.

BrunoGrandePhD replied:
For what it's worth: 

```python
>>> empty = []
>>> ";".join(empty)
''
```

BWMac commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+    with open(input) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            output_dict.update(
```
Yes, I also thought it would be the most efficient way to create the structure I was after.

BWMac commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+import json
+
+
+def clean_json_list(json_list: str) -> list:
+    """
+    Cleans comma-separated string of JSON file paths and produces python list
+
+    Args:
+        json_list (str): comma-separated string with JSON file paths generated from nextflow channels
+
+    Returns:
+        list: python list of JSON file paths with validation test results
+    """
+    lst = json_list.replace("[", "").replace("]", "").split(",")
+    final_list = [item.strip() for item in lst if item.endswith(".json")]
+    final_list = list(set(final_list))
+    return final_list
```
I believe that I still need this function, as the list of JSON files still comes in as a comma-separated string rather than a list.

BrunoGrandePhD commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+    json_list = sys.argv[1]
+    input = sys.argv[2]
+    # clean up json_list
+    final_list = clean_json_list(json_list)
```
This function was eliminated.

BrunoGrandePhD commented on bin/csv_output.py:
```diff
@@ -0,0 +1,149 @@
+        list: list of dictionaries containing all data needed for output csv file
+    """
+    output_list = []
+    for k, v in output_dict.items():
```
Now that I'm looking at this, I might delete this function in favor of a single line wherever it's being called:

```python
output_list = list(output_dict.values())

# OR

output_list = [*output_dict.values()]
```

thomasyu888 replied:
Til: That^.  Great suggestion.

BWMac commented:
The latest commit addresses comments and fixes the issue that we were having running the pipeline on Tower. When workflows run on Tower, they cannot access the `assets` path in worker jobs, and this was causing both the JSON files and the input csv file to be unreadable by the CSV_OUTPUT process. (Thanks @BrunoGrandePhD for your help figuring this out). The workflow has been successfully tested on Tower.

BrunoGrandePhD commented on bin/csv_output.py:
```diff
@@ -0,0 +1,126 @@
+                output_dict[syn_id]["failed_tests"].append(test_name)
+
+    # set is_valid
+    for k in output_dict:
+        if set(output_dict[k]["passed_tests"]).issuperset(
+            set(output_dict[k]["required_tests"])
+        ):
+            is_valid = True
+        else:
+            is_valid = False
+        output_dict[k]["is_valid"] = is_valid
```
**Nit (_i.e._ [optional](https://google.github.io/eng-practices/review/reviewer/standard.html#:~:text=Reviewers%20should%20always%20feel%20free,they%20could%20choose%20to%20ignore.)):** By using the `.items()` dict method, you can streamline this for-loop and if-else statement as follows:

```suggestion
    for k, v in output_dict.items():
        v["is_valid"] = set(v["passed_tests"]) >= set(v["required_tests"])
```

BrunoGrandePhD commented on bin/csv_output.py:
```diff
@@ -0,0 +1,126 @@
+
+    # export to CSV
+    keys = output_list[0].keys()
+    with open("validation_results.csv", "w", newline="") as output_file:
```
This is the first time seeing `newline=` being used in `open()`. Could you add a comment explaining why this is needed.

BWMac replied:
I have always done it this way, I guess it is just how I learned it. There is some sound rationale in this [discussion board](https://discuss.codecademy.com/t/what-does-the-newline-argument-do/463575). And I have added a code comment as well.

BrunoGrandePhD reviewed (approved):
One nit and one request for a code comment, but I'll approve now. 🚀

thomasyu888 reviewed (approved):
LGTM! Thanks Brad!
